### PR TITLE
WIP: Propagate block parts rarest first

### PR DESCRIPTION
**Pull request by ebuchman**
Still need to write tests

### CODE REVIEW

**Comment by melekes** [comment]
What was the reason to do this?

**Comment by ebuchman** [comment]
Performance. I originally wrote this code while running thesis experiments last year. I couldn't really see a performance improvement at the time, but with recent testnet work thought I would resurrect the code and revisit. In theory it should help

File: types/part_set.go [hunk 1]
```diff
@@ -274,3 +279,74 @@ func (ps *PartSet) StringShort() string {
...
+	defer pc.mtx.Unlock()
+	pc.counts[index] += 1
+}
+
+// Pick the first index from the bit array with the lowest count, increment the index,
+// and pass it to the setFunc.
+// Index must be incremented so that multiple concurrent calls don't pick the same value,
+// and passed to setFunc so the abundance doesnt get accidentally incremented again
+// Assumes possibleParts.Size() == len(pc.counts)
+func (pc *PartCounts) PickRarest(possibleParts *cmn.BitArray, setFunc func(int)) int {
```
Comment (melekes) [inline types/part_set.go]: can't we use here some tree (like binary heap) where the rarest element would always be on top?

File: consensus/reactor.go [hunk 2]
```diff
@@ -425,14 +461,21 @@ OUTER_LOOP:
 
 		// Send proposal Block parts?
 		if rs.ProposalBlockParts.HasHeader(prs.ProposalBlockPartsHeader) {
-			if index, ok := rs.ProposalBlockParts.BitArray().Sub(prs.ProposalBlockParts.Copy()).PickRandom(); ok {
+			possibleParts := rs.ProposalBlockParts.BitArray().Sub(prs.ProposalBlockParts.Copy())
+			if !possibleParts.IsEmpty() {
+				index := rs.ProposalBlockParts.Abundance.PickRarest(possibleParts, func(setIndex int) {
```
Comment (melekes) [inline consensus/reactor.go]: why pick func need to accept `setFunc`? can't we call setFunc later if the `Send` succeeds?

**Comment by odeke-em** [comment]
@melekes this is the issue that describes its utility https://github.com/tendermint/tendermint/issues/204.
@ebuchman perhaps we should also tag this in the body of the commit message as
`Fixes #204` and also in this issue's description?

**Comment by odeke-em** [comment]
Also seems like we need to fix a merge conflict

File: consensus/reactor.go (hunk 2 above)
Comment (odeke-em) [reply consensus/reactor.go]: Yeah, @ebuchman puts a TODO comment in the body of this code wondering what should we do if send fails.

**Comment by odeke-em** [comment]
@ebuchman aside from the tests, we need to figure out a way to benchmark/measure our throughput and overall speed.

File: types/part_set.go [hunk 3]
```diff
@@ -274,3 +279,74 @@ func (ps *PartSet) StringShort() string {
...
+	if check() {
+		pc.counts[index] += 1
+	}
+}
+
+// panics if index >= len(pc.counts)
+func (pc *PartCounts) IncrementIndex(index int) {
+	pc.mtx.Lock()
+	defer pc.mtx.Unlock()
+	pc.counts[index] += 1
```
Comment (cloudhead) [inline types/part_set.go]: Can we use `atomic.AddUInt32` here and get rid of the lock?

Comment (ebuchman) [reply types/part_set.go]: Believe so - this code can probably be improved a bunch

File: types/part_set.go (hunk 1 above)
Comment (ebuchman) [reply types/part_set.go]: yep believe so!

**Comment by ebuchman** [comment]
closing for issue https://github.com/tendermint/tendermint/issues/204

**Pull request closed without merging**